processor/src/test/resources/fixtures/org/mapstruct/ap/test/nestedbeans/mixed/FishTankMapperImpl.java

```diff
@@ -0,0 +1,358 @@
+        }
+
+        WaterQualityOrganisationDto waterQualityOrganisationDto = new WaterQualityOrganisationDto();
+
+        waterQualityOrganisationDto.setName( waterQualityReport.getOrganisationName() );
+
+        return waterQualityOrganisationDto;
+    }
+
+    protected WaterQualityReportDto waterQualityReportToWaterQualityReportDto(WaterQualityReport waterQualityReport) {
+        if ( waterQualityReport == null ) {
+            return null;
+        }
+
+        WaterQualityReportDto waterQualityReportDto = new WaterQualityReportDto();
+
+        waterQualityReportDto.setOrganisation( waterQualityReportToWaterQualityOrganisationDto( waterQualityReport ) );
+        waterQualityReportDto.setVerdict( waterQualityReport.getVerdict() );
+
+        return waterQualityReportDto;
+    }
+
+    protected WaterQualityDto waterQualityToWaterQualityDto(WaterQuality waterQuality) {
+        if ( waterQuality == null ) {
+            return null;
+        }
+
+        WaterQualityDto waterQualityDto = new WaterQualityDto();
+
+        waterQualityDto.setReport( waterQualityReportToWaterQualityReportDto( waterQuality.getReport() ) );
+
+        return waterQualityDto;
+    }
+
+    private Ornament sourceInteriorOrnament(FishTank fishTank) {
+
+        if ( fishTank == null ) {
+            return null;
+        }
+        Interior interior = fishTank.getInterior();
+        if ( interior == null ) {
+            return null;
+        }
+        Ornament ornament = interior.getOrnament();
+        if ( ornament == null ) {
+            return null;
+        }
+        return ornament;
+    }
+
+    protected OrnamentDto ornamentToOrnamentDto(Ornament ornament) {
+        if ( ornament == null ) {
+            return null;
+        }
+
+        OrnamentDto ornamentDto = new OrnamentDto();
+
+        ornamentDto.setType( ornament.getType() );
+
+        return ornamentDto;
+    }
+
+    protected WaterPlantDto waterPlantToWaterPlantDto(WaterPlant waterPlant) {
+        if ( waterPlant == null ) {
+            return null;
+        }
+
+        WaterPlantDto waterPlantDto = new WaterPlantDto();
+
+        waterPlantDto.setKind( waterPlant.getKind() );
+
+        return waterPlantDto;
+    }
+
+    private Ornament sourceInteriorOrnament1(FishTank fishTank) {
+
+        if ( fishTank == null ) {
+            return null;
+        }
+        Interior interior = fishTank.getInterior();
+        if ( interior == null ) {
+            return null;
+        }
+        Ornament ornament = interior.getOrnament();
+        if ( ornament == null ) {
+            return null;
+        }
+        return ornament;
+    }
+
+    protected Fish fishDtoToFish(FishDto fishDto) {
+        if ( fishDto == null ) {
+            return null;
+        }
+
+        Fish fish = new Fish();
+
+        fish.setType( fishDto.getKind() );
+
+        return fish;
+    }
+
+    private String waterQualityReportDtoOrganisationName(WaterQualityReportDto waterQualityReportDto) {
```

Comment:
totally agree. Dunno how difficult it is to handle. I think the same logic is re-used. I think that now this is handled in `AbastractMehthodBuilderBase#String getName(Type sourceType, Type targetType)`.. Perhaps we should add an enum and pass that along the call stack: e.g. MethodStyle (GETTER, MAPPER)..